### DISCUSSION TITLE
Drift when dynamic object passes

### DISCUSSION THREAD

**Discussion by stevemartinov** [stella-cv/stella_vslam#228] [Q&A] [answered]
Hi,

I have faced a problem when mapping. The big object (sometimes even people) do pass in front of the camera while the camera is stationary and features are being extracted from those objects. Then, the camera drifts a lot. What would be a good strategy here? I do not think that IMU integration would help because after checking your IMU branch, it still takes a significant information from the extracted keypoints and IMU is not being fed into a filter.

**Comment by ymd-stella** [comment, marked as answer]
IMU integration is not yet complete. There are no plans to proceed with the development of IMU integration at this time.

I recommend using a camera with a wide field of view (fisheye camera) or a 360 degree panoramic camera.

**Reply by stevemartinov** [reply]
I did test the IMU branch and it does seem to work especially with a scale issue. What is not complete that really stops it from using this feature?

**Reply by ymd-stella** [reply]
Everything is unimplemented except for the data definition and preprocessing. For example, optimization.

**Reply by stevemartinov** [reply]
Ok, I will start implementing. Is there any specific paper I can follow to integrate the IMU?

**Reply by ymd-stella** [reply]
See [On-Manifold Preintegration for Real-Time Visual-Inertial Odometry](https://arxiv.org/abs/1512.02363) and [Inertial-Only Optimization for Visual-Inertial Initialization](https://arxiv.org/abs/2003.05766).

There is an experimental implementation, but it has some issues.
https://github.com/ymd-stella/openvslam-community/tree/wip/gba.

- The edge implementation is untested and suspicious.
- It pollutes the existing global BA.
- It is very unstable. The result will change each time you run it. In particular, the calculation of the direction of gravity is unstable.
- There seems to be a difference in the optimal strategy for keyframe insertion: compared to Visual SLAM, we needed to insert more keyframes for scale and gravity direction initialization.
- My impression from the experiments so far is that the time taken for initialization is too long. In some applications, it may be necessary to move the camera randomly until it is initialized.

**Reply by ymd-stella** [reply]
For your purpose, you need to implement only prediction, and may not need initialization and graph optimization (if the gravity component is removed from the IMU input by preprocessing).

**Comment by an99990** [comment]
would it be possible to filter features coming from moving objects using optical flow ? moving features would be eliminated ? This is a feature that I wan to work on, any thoughts ?

**Reply by ymd-stella** [reply]
I think the problem is poorly defined. You say remove moving objects, but if you are in a train, what is the moving object? Do you remove the train?

Why would you want to remove the moving object? Do you really need to remove the moving object?

**Reply by an99990** [reply]
What Id like to do is remove feature belonging to moving objects so it doesnt affect the pose estimation of the camera. like ds-slam is doing: https://arxiv.org/pdf/1809.08379.pdf . features extracted from a person walking by would be removed from the calculation.

**Reply by an99990** [reply]
![image](https://user-images.githubusercontent.com/87030365/171685204-c3cbc0d7-4acc-450a-a600-3810d7373b41.png)

from their paper

**Reply by ymd-stella** [reply]
> What Id like to do is remove feature belonging to moving objects so it doesnt affect the pose estimation of the camera.

Have you checked to see what moving objects affect the pose estimation?

If the camera's field of view is wide enough, moving object is ignored by pose optimization or local BA. Then, landmarks that cannot be tracked are removed by local_map_cleaner.

**Reply by ymd-stella** [reply]
@an99990 
Please do not add your question where another person has asked something unless it is the exact same issue. It will send an unwanted notification to the first questioner. Please create a new question and include a reference link.
It will also be harder to get answers because we will not be able to track unanswered questions.
